crates/node/src/roles/worker.rs

```diff
@@ -86,73 +85,140 @@ impl WorkerRole {
-                result.map_err(WorkerRoleError::AdminPanic)??;
-                panic!("Unexpected termination of admin.");
-            },
-            result = &mut worker_handle => {
-                result.map_err(WorkerRoleError::WorkerPanic)??;
-                panic!("Unexpected termination of worker.");
+            Some(component_result) = component_set.join_next() => {
+                let component_name = component_result.map_err(WorkerRoleError::ComponentPanic)??;
+                panic!("Unexpected termination of component '{component_name}'");
+            }
+            bifrost_result = &mut bifrost_join_handle => {
+                bifrost_result.map_err(WorkerRoleError::ComponentPanic)??;
+                panic!("Unexpected termination of bifrost service");
             }
-            result = &mut bifrost_handle => {
-                result.map_err(WorkerRoleError::BifrostPanic)??;
-                panic!("Unexpected termination of bifrost service.");
-            },
         }
 
         Ok(())
     }
+
+    async fn reload_schemas<SC>(
+        subscription_controller: Option<SC>,
+        schemas: Schemas,
+        schema_channel: Channel,
+    ) -> Result<(), WorkerRoleError>
+    where
+        SC: SubscriptionController + Clone + Send + Sync,
+    {
+        // todo: make this configurable
+        let mut fetch_interval = tokio::time::interval(Duration::from_secs(5));
+        let mut schema_grpc_client =
+            restate_node_services::schema::schema_client::SchemaClient::new(schema_channel);
+
+        loop {
+            fetch_interval.tick().await;
+
+            debug!("Trying to fetch schema information");
+
+            let result = schema_grpc_client
+                // todo introduce schema version information to avoid fetching and overwriting the schema information
+                //  over and over again
+                .fetch_schemas(FetchSchemasRequest {})
+                .await;
+            match result {
+                Ok(response) => {
+                    let (schema_updates, _) =
+                        bincode::serde::decode_from_slice::<Vec<SchemasUpdateCommand>, _>(
+                            &response.into_inner().schemas,
+                            bincode::config::standard(),
+                        )?;
+
+                    // hack to suppress repeated logging of schema registrations
+                    // todo: Fix it
+                    tracing::subscriber::with_default(NoSubscriber::new(), || {
+                        schemas.overwrite(schema_updates)
+                    })?;
+
+                    if let Some(subscription_controller) = subscription_controller.as_ref() {
+                        Self::update_subscriptions(&schemas, subscription_controller.clone())
+                            .await?;
+                    }
+                }
+                Err(err) => {
+                    debug!("Failed fetching schema information: {err}")
```

Comment:
Should this be a debug in the single-node setup? Why do we expect failures to happen?

Reply:
In the single-node setup this should indeed not happen. It is a bit of a preparational step for allowing the cluster controller and the worker role to be run as two separate processes.

Reply:
I'd rather make this scream loud if we are confident that it shouldn't happen :)

Reply:
By allowing to separate the admin from the worker role which will be the next step I intend to implement, it will be possible that the worker cannot reach the admin.